### PR TITLE
Spots visualizer

### PR DESCRIPTION
**Pull request by tddough98**
Here we add a variant of DeepCell Label client to display the spots consumer output and rework the backend to load data into a standardized archive to be loaded by the client.

The visualizer is online at [spots.deepcell.org](https://spots.deepcell.org). You can create a project by sending a form with two URLs. The form should have a `images` field with a download URL for a file with the raw image data and an optional `labels` field with a download URL for a zip of the labels. The spots.deepcell.org address hosts both the client and the server. Here's an example CURL command to create a project (links may expire): 
```
curl --location --request POST 'spots.deepcell.org/api/project' \
--form 'images="https://storage.googleapis.com/deepcell-prod/uploads/2ebc4a3339b0fc260b8c4815a47888fd.png"' \
--form 'labels="https://drive.google.com/uc?export=download&id=1tRCvgcs_1MzZzVGJukgu1pQr-m2vrTs5"'
```

The request will return a project ID like `l5lCB8ervg7C` that can be viewed at [spots.deepcell.org/project?projectId=l5lCB8ervg7C](https://spots.deepcell.org/project?projectId=l5lCB8ervg7C).

Here are the changes to visualize spots on the client:
 - Add a SpotsCanvas to display spots 
 - Add a spotsMachine to manage the spots data and display controls
 - Add SpotsControls to change how spots are displayed
 - Use a REACT_APP_SPOTS_VISUALIZER environment variable feature flag so that the complete labeling app and the simplified spots viewer exist in the same code base
 
These are other significant changes to support spots:
- Make other controls more compact
- Create a new simplified backend that generates a zip from input files, uploads it to an S3 bucket, and serves the files to the client
- Introduce a new loadMachine on the client to read this zip

There are also changes to codebase quality included in the mix like:
 - Add index.js to component folders to simplify the imports (e.g. changing Footer/Footer to Footer)
 - Rename QualityControl___ components to Review___
 - Combine useSelector hooks into one hook that compares results with `equal` from 'fast-deep-equal'
 - Remove a intermediate drawing canvas from components that don't need them
 - Separate components defined in App.js into their own files

As there are many changed files, here's a guide to the relevant files for each changes, excluding trivial changes (renaming, moving a file).

- backend rework: bleuprints.py, models.py, loaders.py, labelmaker.py
- testing new backend conftest.py, blueprints_test.py, models_test.py, loaders_test.py, labelmaker_test.py
- end to end tests for Label: label.spec.js, rgb.zip, tests.yaml
- e2e tests for Spots: spots.spec.js, spots.zip, tests.yaml
- loading spots on client: package.json, loadMachine.ts, spotsMachine.js
- displaying spots: SpotsCanvas.js, ProjectContext.js, canvasMachine.js, 
- controlling spots display: SpotColorToggle.js, SpotOpacitySlider.js, SpotOutlineToggle.js, SpotRadiusSlider.js, SpotsCheckbox.js, SpotsControls.js
- refactoring App.js: InvalidId.js, Display.js, Load.js, LoadReview.js, Project.js,

### CODE REVIEW

**Review by msschwartz21 — COMMENTED** [review]
Can you clarify what's happening with some of the tracking functionality? I suspect that I'm just losing track of code when it's being rearranged, but I do want to confirm that we are not losing any functionality unintentionally.

File: deepcell_label/exporters.py [hunk 1]
```diff
@@ -1,126 +0,0 @@
-"""Classes to export a DeepCell Label project as a .npz or .trk file."""
```
[the remark this answers is not in the document]
Comment (tddough98) [reply deepcell_label/exporters.py:1]: Export is currently missing from the dev branch.

For the visualizers, we can download the initial project zip as we don't make any changes to the data. For editing projects, I'm going to work a new route that takes data from the clients and repackages it into a zip with the same contents/format as the input zip.

File: deepcell_label/loaders.py [hunk 2]
```diff
@@ -1,357 +1,342 @@
...
-        elif is_png(self.labeled_path):
-            label_array = load_png(data)
-        elif is_tiff(self.labeled_path):
-            label_array = load_tiff(data)
-        elif is_zip(self.labeled_path):
-            label_array = load_zip(data)
-        else:
-            ext = pathlib.Path(self.labeled_path).suffix
-            raise InvalidExtension('invalid file extension: {}'.format(ext))
+    def __init__(self, image_file=None, label_file=None):
```
[the remark this answers is not in the document]
Comment (tddough98) [reply deepcell_label/loaders.py:24]: These are actually `tempfile.TemporaryFile` objects. We request the data from the URLs and load them into the temporary files in the `/api/project` route before creating the Loader object.

File: deepcell_label/loaders.py [hunk 3]
```diff
@@ -1,357 +1,342 @@
...
+        f.seek(0)
+        X = TiffFile(io.BytesIO(f.read())).asarray(squeeze=False)
+        if X.ndim == 2:
+            # Add channel and frame axes
+            X = X[np.newaxis, ..., np.newaxis]
+        if X.ndim == 3:
+            # TODO: use dimension order to know whether to add frame or channel axis
+            X = X[np.newaxis, ...]
+        if X.ndim > 4:
+            raise ValueError('Tiff file has more than 4 dimensions')
```
[the remark this answers is not in the document]
Comment (tddough98) [reply deepcell_label/loaders.py:314]: You're right, these should be more general and abstracted. As we add in the dimension order for batches and channels in #319, it'll be a good time to find common patterns in reshaping. For this visualizers, the inputs are more limited, so I think we're fine for now, but definitely something to keep an eye on.

File: deepcell_label/loaders.py [hunk 4]
```diff
@@ -1,357 +1,342 @@
...
-    with tempfile.NamedTemporaryFile() as temp:
-        temp.write(data)
-        with tarfile.open(temp.name, 'r') as trks:
-            with io.BytesIO() as array_file:
-                array_file.write(trks.extractfile('raw.npy').read())
-                array_file.seek(0)
-                return np.load(array_file)
-
-
-def load_labeled_trk(data):
```
[the remark this answers is not in the document]
Comment (tddough98) [reply deepcell_label/loaders.py:270]: I'm moving away from using .trk files anywhere in DCL. In #322, I recreated the .trk example files as .zip instead. We could add in .trk loading functionality if we need it for jobs, but switching to zips matches the direction that DCL is headed in

File: deepcell_label/blueprints.py [hunk 5]
```diff
@@ -168,98 +164,6 @@ def edit(action):
     return response
 
 
-@bp.route('/api/undo/<token>', methods=['POST'])
```
[the remark this answers is not in the document]
Comment (tddough98) [reply deepcell_label/blueprints.py:171]: Undo/redo is the biggest regression on the dev branch right now and highest on my priorities.

Editing and storage are now completely decoupled. We edit labels by sending a zip file of labels to the edit route, which responds with a zip of the edited labels. If we store both the outgoing and incoming zips, we'll be able to load the labels before and after the edit to undo or redo it.

The overlaps branch implements the zip exchange between the client and the server, so by merging that in, I'll be a step closer to implementing undo redo on top of that. I'll create a PR soon that adds in overlapping labels and hopefully fix the lineage editing that is not yet working in #322.

**Comment by tddough98** [comment]
We're starting to build up more PRs like #322 on top of this one, so I will merge this in now. I addressed @msschwartz21 comments, and it doesn't seem like there are code changes needed yet, but rather future PRs to address regressions needed to bring the editing app back into production. The spots visualizer is set to come online and the editing app doesn't have any further regressions from this PR.

**Pull request merged**